File: HashMap/383 - Ransom Note/RansomNote.cpp
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {

        // Alternate implementation 
        
        // unordered_map<char, int> magMap;

        // for (int i = 0; i < magazine.length(); i++) {
        //     if (magMap.find(magazine[i]) != magMap.end()) {
        //         magMap[magazine[i]] += 1;
        //     } else {
        //         magMap[magazine[i]] = 1;
        //     }
        // }

        // for (int j = 0; j < ransomNote.length(); j++) {
        //     if (magMap.find(ransomNote[j]) != magMap.end()) {
        //         if (magMap[ransomNote[j]] > 1) {
        //             magMap[ransomNote[j]] -= 1;
        //         } else {
        //             magMap.erase(ransomNote[j]);
        //         }
        //     } else {
        //         return false;
        //     }
        // }

        for (int i = 0; i < ransomNote.length(); i++) {
            // Tries to search for character, if successful, returns index, else
            // returns -1.
            int index = magazine.find(ransomNote[i]);
            if (index != -1) {
                magazine.erase(index, 1);   // Index to erase and the count of characters to erase
            } else {
                return false;
            }
        }

        return true;
    }
};

File: HashMap/383 - Ransom Note/RansomNote.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        
        magDict = {}    # Create dict to store characters of magazine string

        # Store the characters of the magazine string
        for i in magazine:
            if (magDict.get(i) != None):
                # Increment Value of the key by 1 if present already
                magDict[i] += 1
            else:
                # Add a new key-value pair
                magDict[i] = 1
        

        # Checks if the characters in the note are present in the dict
        for j in ransomNote:
            if (magDict.get(j) != None):
                # if the character is present and their value is >1, then
                # decrement the value by 1
                if (magDict.get(j) > 1):
                    magDict[j] -= 1
                # If the character is present and their value is =1, then remove
                # the key-value pair from dict
                else:
                    # Pop the character. If the character is not present, then
                    # return None
                    magDict.pop(j, None)  
            # If character is not present then return False
            else:
                return False
        
        return True
        
# Alternate implementation below
"""
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        for letter in ransomNote:
            if magazine.find(letter) != -1:
                magazine = magazine.replace(letter, '', 1)
            else:
                return False

        return True
"""
    